### PR TITLE
feat(dotnet): support local debug for DotNet

### PR DESCRIPTION
**Pull request by hund030**
E2E TEST: cli does not support DotNet scenario yet.

When trigger local debug in VS platform, setup local setting and update appsetting.development.json file in .NET project.

### CODE REVIEW

**Review by xiaolang124 — APPROVED** [review]

**Review by JerryYangKai — APPROVED** [review]

**Review by swatDong — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts [hunk 1]
```diff
@@ -84,3 +96,11 @@ export class WebappBicep {
     domainAsParam: WebappBicep.domainAsParam,
   };
 }
+
+export class AppSettingsPlaceholder {
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts:100]: Nit: using plural form may be better - `AppSettingsPlaceholders`

File: packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts [hunk 2]
```diff
@@ -84,3 +96,11 @@ export class WebappBicep {
...
+  static readonly clientId = "$clientId$";
+  static readonly clientSecret = "$client-secret$";
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts:102]: Nit: Naming conventions are inconsistent.
`clientId`, `botId` vs `client-secret`, `bot-password`

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts [hunk 3]
```diff
@@ -202,6 +212,66 @@ export class DotnetPluginImpl implements PluginImpl {
...
+      ctx.envInfo.state.get(PluginInfo.pluginName)?.set(ConfigInfo.indexPath, PathInfo.indexPath);
+    } else {
+      ctx.localSettings?.frontend?.set(LocalSettingsFrontendKeys.TabIndexPath, PathInfo.indexPath);
+    }
+    return ok(undefined);
+  }
+
+  public async postLocalDebug(ctx: PluginContext): Promise<TeamsFxResult> {
+    const appSettingsPath = path.join(ctx.root, PathInfo.appSettingDevelopment);
+    let appSettings = await fs.readFile(appSettingsPath, "utf-8");
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:225]: To be robust, check if the file is existing or catch related exceptions.

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts [hunk 4]
```diff
@@ -202,6 +212,66 @@ export class DotnetPluginImpl implements PluginImpl {
...
+        ? PathInfo.oauthHost(tenantId)
+        : AppSettingsPlaceholder.oauthAuthority;
+      botId =
+        ctx.localSettings?.bot?.get(LocalSettingsBotKeys.BotId) ?? AppSettingsPlaceholder.botId;
+      botPassword =
+        ctx.localSettings?.bot?.get(LocalSettingsBotKeys.BotPassword) ??
+        AppSettingsPlaceholder.botPassword;
+    }
+
+    appSettings = appSettings.replace(AppSettingsPlaceholder.clientId, clientId);
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:269]: Nit: Using template engine like mustache may be better?

File: packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts [hunk 5]
```diff
@@ -143,6 +143,24 @@ export async function setupLocalDebugSettings(
...
+      if (includeBot) {
+        localSettings.bot ??= {};
+        const ngrokHttpUrl = await getNgrokHttpUrl(2544);
+        if (!ngrokHttpUrl) {
+          const error = NgrokTunnelNotConnected();
+          TelemetryUtils.sendErrorEvent(TelemetryEventName.setupLocalDebugSettings, error);
+          return err(error);
+        }
+        localSettings.bot.botEndpoint = ngrokHttpUrl;
+        localSettings.bot.botDomain = ngrokHttpUrl.slice(8);
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts:162]: Nit: Kind of not robust - better to leverage some library to extract domain from url.

**Review by IvanJobs — APPROVED** [review]
lgtm - assumed that dev test is passed.

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts [hunk 6]
```diff
@@ -202,6 +212,66 @@ export class DotnetPluginImpl implements PluginImpl {
...
+      botPassword =
+        ctx.envInfo.state.get(PluginNames.BOT)?.get(DependentPluginInfo.botPassword) ??
+        AppSettingsPlaceholder.botPassword;
+    } else {
+      clientId =
+        ctx.localSettings?.auth?.get(LocalSettingsAuthKeys.ClientId) ??
+        AppSettingsPlaceholder.clientId;
+      clientSecret =
+        ctx.localSettings?.auth?.get(LocalSettingsAuthKeys.ClientSecret) ??
+        AppSettingsPlaceholder.clientSecret;
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:257]: Nit: these lines are kind of verbose and within similar biz logic. Better to simplify them to improve readability if time allows.

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts (hunk 3 above)
Comment (hund030) [reply packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:225]: Let's fix it in next PR.

File: packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts [hunk 7]
```diff
@@ -38,6 +38,11 @@ export class DotnetPathInfo {
     path.join(templateFolder, "plugins", "resource", "webapp", "bicep");
   static readonly TemplateFolderName = "dotnet";
   static readonly projectFilename = (projectName: string): string => `${projectName}.csproj`;
+
+  static readonly indexPath = ""; // Index path is '/', relational path is empty.
```
Comment (eriolchan) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts:42]: typo: relational => relative

File: packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts (hunk 2 above)
Comment (eriolchan) [reply packages/fx-core/src/plugins/resource/frontend/dotnet/constants.ts:102]: Also curious why the naming conventions are different?

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts [hunk 8]
```diff
@@ -202,6 +212,66 @@ export class DotnetPluginImpl implements PluginImpl {
   }
 
   public async localDebug(ctx: PluginContext): Promise<TeamsFxResult> {
+    if (isConfigUnifyEnabled()) {
```
Comment (eriolchan) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:215]: Curious on what is this flag "isConfigUnifyEnabled1" for?

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts (hunk 4 above)
Comment (eriolchan) [reply packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:269]: Is there any possible for the case of replaceAll()?

File: packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts [hunk 9]
```diff
@@ -143,6 +143,24 @@ export async function setupLocalDebugSettings(
...
+    } else if (inputs.platform === Platform.VS) {
+      if (includeFrontend) {
+        localSettings.frontend ??= {};
+        localSettings.frontend.tabEndpoint = "https://localhost:44302";
+        localSettings.frontend.tabDomain = "localhost";
+      }
+
+      if (includeBot) {
+        localSettings.bot ??= {};
+        const ngrokHttpUrl = await getNgrokHttpUrl(2544);
```
Comment (eriolchan) [inline packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts:155]: Does 2544 refer to port number? It looks like a magic number

File: packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts (hunk 5 above)
Comment (eriolchan) [reply packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts:162]: Maybe we can add comment here to explain why use 8 as magic number

File: packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts [hunk 10]
```diff
@@ -253,6 +271,25 @@ export async function setupLocalEnvironment(
           }
         }
       }
+    } else if (inputs.platform === Platform.VS) {
```
Comment (eriolchan) [inline packages/fx-core/src/plugins/solution/fx-solution/debug/provisionLocal.ts:274]: Looks like the similar logic as above, is that possible to refactor it in the future?

**Pull request merged**